core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java

```diff
@@ -561,6 +593,234 @@ public UpdateNode plan(Update node)
+            projectionAssignmentsBuilder.put(symbol, target.getFieldMappings().get(fieldIndex).toSymbolReference());
+        }
+
+        FieldReference reference = analysis.getRowIdField(mergeAnalysis.getTargetTable());
+        Field rowIdField = target.getDescriptor().getFieldByIndex(reference.getFieldIndex());
+        Symbol rowId = symbolAllocator.newSymbol("row_id", rowIdField.getType());
+        projectionAssignmentsBuilder.put(rowId, target.getFieldMappings().get(reference.getFieldIndex()).toSymbolReference());
+        Symbol mergeOutput = symbolAllocator.newSymbol("merge_row", rowType);
+        projectionAssignmentsBuilder.put(mergeOutput, castCaseExpression);
+
+        ProjectNode project = new ProjectNode(
+                idAllocator.getNextId(),
+                subPlan.getRoot(),
+                projectionAssignmentsBuilder.build());
+
+        // Projecting the writeRedistributionColumns, the rowId block, and the merge RowBlock
+        int expectedSize = redistributionColumns.size() + 1 + 1;
+        int actualSize = project.getOutputSymbols().size();
+        checkArgument(actualSize == expectedSize, "projectedSymbols should have size %s, but is %s", expectedSize, actualSize);
+
+        Table table = merge.getTarget();
+        TableHandle handle = analysis.getTableHandle(table);
+        TableMetadata tableMetadata = metadata.getTableMetadata(session, handle);
+
+        RowChangeParadigm paradigm = metadata.getRowChangeParadigm(session, handle);
+        Type rowIdType = analysis.getType(analysis.getRowIdField(table));
+        RowChangeProcessor rowChangeProcessor = createMergeProcessor(paradigm, tableMetadata, mergeAnalysis.getDataColumnHandles(), mergeAnalysis.getRedistributionColumnHandles(), rowIdType);
+        MergeTarget mergeTarget = new MergeTarget(handle, Optional.empty(), tableMetadata.getTable(), mergeDetails, rowChangeProcessor);
+
+        ImmutableList.Builder<Symbol> columnSymbolsBuilder = ImmutableList.builder();
+        for (ColumnHandle columnHandle : mergeAnalysis.getDataColumnHandles()) {
+            int fieldIndex = requireNonNull(mergeAnalysis.getColumnHandleFieldNumbers().get(columnHandle), "Could not find field number for column handle");
+            columnSymbolsBuilder.add(target.getFieldMappings().get(fieldIndex));
+        }
+        List<Symbol> columnSymbols = columnSymbolsBuilder.build();
+
+        ImmutableList.Builder<Symbol> projectedSymbolsBuilder = ImmutableList.builder();
+        int subscriptIndex = 1;
+        for (Symbol columnSymbol : columnSymbols) {
+            SubscriptExpression subscriptExpression = new SubscriptExpression(new SymbolReference(mergeOutput.getName()), new GenericLiteral("BIGINT", String.valueOf(subscriptIndex)));
+            projectedSymbolsBuilder.add(columnSymbol);
+            Type type = mergeAnalysis.getDataColumnSchemas().get(subscriptIndex - 1).getType();
+            analysis.addTypes(ImmutableMap.of(NodeRef.of(subscriptExpression), type));
```

Comment:
Why do you need to record the types of those `SubscriptExpression`s? We should try to avoid adding things to the Analysis at this stage.

Reply:
This whole if statement is no longer needed, and I've removed it.  It was a leftover from the time I did tree-building in the `StatementAnalyzer`.